### PR TITLE
HPC-9034 Update Nx 16, React 18 and respective libraries

### PR DESCRIPTION
**Pull request by Onitoxan**
## Describe your changes:

- 🚀 Upgraded Nx to v16.3.1
- 🚀 Upgraded React to v18.2.0
- 🚀 Upgraded Jest to v29.5.0 (Nx upgrade related)
- 🚀 Upgraded React Router to v6.11.2
- 📦  Different package updates
- Updated `<Switch>` React Router v5 component to `<Routes>` React Router v6
- Updated `<Route>` structure, React Router v6 has been thought to try to use relative paths (Didn't want to refactor too much to not go out of the task, but we could define how to write it correctly and that is easy to understand)
- Added `usePrompt()` hook to show prompts in the application when we need
 
Please refer to the commit messages to check a more in-detailed explanation of the changes step by step

## How can I test the changes?:

You should init the app and test [here](http://localhost:4200)
To test the changes is important to:

1. Check the navigation through the page and check that no link has been broken
2. Check that in the Enketo form of any operation, you can't leave an unsaved form without a prompt appearing (For reload action and closing tab action, the browser prompt should appear instead of ours)
3. The application loads correctly
4. Check package.json to see if something seems off

## TO DO:
I would say there is nothing left to do of the task

## Does any part of this code spark :star: joy :star:?:

No special or remarkable things were done I would say. (Except becoming a full-time reader of documentation)

### CODE REVIEW

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

File: apps/hpc-cdm/src/app/components/enketo/index.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/components/enketo/index.tsx]: Don't comment out code, delete it.

File: apps/hpc-cdm/src/app/pages/operations-list.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/pages/operations-list.tsx]: No need to remove empty lines

File: apps/hpc-cdm/src/app/pages/operation.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/pages/operation.tsx:53]: Good! We don't want any dead code lying around

**Review by Pl217 — CHANGES_REQUESTED** [review]
Please split the first commit where you upgrade from Nx 13 directly to Nx 15. Even though Nx migration scripts upgrade all at once, you don't have to commit all changes all at once. Please see this PR #287.

Also, splitting your changes into individual smaller commits helps me understand how these changes are made and makes the review tremendously easier.

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

File: package.json
Comment (Pl217) [inline package.json]: Downgrading Babel and MUI libraries is very likely due to bad conflict resolution when rebasing on top of #373. Amend this commit and don't touch these libraries, as they are irrelevant for Nx upgrade.

File: package.json
Comment (Pl217) [inline package.json]: Downgrade of `@types/dotenv` likely came from a bad rebase on top of #373. This package is not needed and will be removed in #376, but you should not touch it anyway.

File: package.json
Comment (Pl217) [inline package.json]: Don't downgrade `lint-staged`, `oidc-client` and `prettier`. The reason behind this downgrade seems to be a bad rebase once #373 got merged.

File: libs/hpc-ui/src/lib/i18n/index.tsx
Comment (Pl217) [inline libs/hpc-ui/src/lib/i18n/index.tsx]: Type assertions are basically developers saying Typescript to shut up, because we think we know better than the TS compiler :) That is usually not the case, so type assertions should be avoided. Thus, lets remove `Date` as possible type  of `params` argument instead of doing this assertion here.

File: apps/hpc-ftsadmin/src/main.tsx
Comment (Pl217) [inline apps/hpc-ftsadmin/src/main.tsx]: In the other comment I am saying bad things about type assertion and how it should be avoided, because TS compiler usually knows better than us. Here, that's not the case, because we (the developers) know better than Typescript. DOM element with ID `'root'` definitely exists and `document.getElementById` should not return `null`.

But still, I like to avoid type assertions, so let's throw an error if root element cannot be found.
```ts
const rootElement = document.getElementById('root');
if (!rootElement) {
  throw new Error('Failed to find the root element');
}
const root = ReactDOM.createRoot(rootElement);
```

File: apps/hpc-cdm/src/main.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/main.tsx]: ditto

File: apps/hpc-cdm/src/app/components/enketo/powered-by-footer.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/components/enketo/powered-by-footer.tsx]: Unfortunately, SVG logo doesn't load
Screenshot_20230509_111146

File: apps/hpc-cdm/src/app/components/form-assignments-list.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/components/form-assignments-list.tsx]: `StyledToolTip` is the unnecessary alias that you kept in this file (defined below). It was just defined to add some padding around the tooltip. The changes you made with all this context and everything just for a few pixels of padding is not worth it. I would just remove all the code related to padding.

Otherwise, if you think some padding is needed, use simple `Tooltip` (not `StyledToolTip`) and add padding using [the sx prop](https://mui.com/system/getting-started/the-sx-prop/):
```tsx
<Tooltip
  componentsProps={{
    tooltip: { sx: { padding: '8px' } },
  }}
  ...
```

File: apps/hpc-cdm/jest.config.ts
Comment (Pl217) [inline apps/hpc-cdm/jest.config.ts]: Why did you disable ESLint in all of these TS config files? I don't see any linting problem in these files.

File: jest.preset.js
Comment (Pl217) [inline jest.preset.js]: Why is this the only Jest config file kept in Javascript?

File: package-lock.json
Comment (Pl217) [inline package-lock.json]: There are plenty of changes on lockfile that you do in various commits, which just looks like a mistake. Since GitHub doesn't properly allow to comment on commits, all will be condensed in this comment:

- Changes to package lockfile shouldn't be made in commit 64d1e9198222d8c45414108dbc82cecf30f9b2d8.
- Entire next commit (0546baf8e061b01a793ea103ec6d741c8bc786b1) should be dropped. Package `eslint-plugin-react-hooks` has already been updated in #373. Just drop this commit during the rebase and install packages properly in commits that do some real updates.
- Commit after that (0f18afd75ae863b3d2143cc24d12129882dbfcaf) is having a duplicated commit message, but doesn't do anything it claims. Just drop it.
- Finally, third commit you need to drop is 48e7f8b3d6df0981aaf42509b495c3173303f012. Looks like you messed up a bit with git history, but it's easy to fix.

File: package.json
Comment (Pl217) [inline package.json]: Downgrade of `babel-plugin-styled-components` looks like a bad conflict resolution after rebase on top of #373.

File: package.json
Comment (Pl217) [inline package.json]: Downgrades of `fp-ts` and `husky` look like a bad conflict resolution after rebase on top of #373.

File: package.json
Comment (Pl217) [inline package.json]: Downgrades of `intl-messageformat` and `io-ts` look like a bad conflict resolution after rebase on top of #373.

File: package.json
Comment (Pl217) [inline package.json]: You downgraded `prettier` in 7d3bbc1e3776ab929bdeabb42acc438db5bfdd35 and then back up in 9f2cb06fb85e935208b60b75b7da3f1abbc44145. Just don't touch it :)

File: package.json
Comment (Pl217) [inline package.json]: You should update React (and related packages) to latest v18 in 7d3bbc1e3776ab929bdeabb42acc438db5bfdd35 and not touch it again in 9f2cb06fb85e935208b60b75b7da3f1abbc44145

File: package.json
Comment (Pl217) [inline package.json]: Downgrade of `styled-components` looks like a bad conflict resolution after rebase on top of https://github.com/UN-OCHA/hpc-cdm/pull/373.

File: package.json
Comment (Pl217) [inline package.json]: Why was `core-js` added as a dependency? Is it really needed? All other dependencies are specified as `devDependencies`, so should `core-js`.

File: package.json
Comment (Pl217) [inline package.json]: When updating, why don't you update to latest available version? v6.4.1 was released 8 months ago. Let's update to latest version available.

I will reuse this comment for another thing. After the commit where you update `react-router-dom`, there are three commits in which you do required migrations. You have divided work in areas where you make changes. That is good. There could be another approach, where you divide work in commits by relevant type of migration. For example, I see that you changed all `<Switch>` elements to `<Routes>`. The commit message could then point to the [exact section](https://reactrouter.com/en/main/upgrading/v5#upgrade-all-switch-elements-to-routes) in migration documentation.

But, you don't have to make this change, it's just how I would divide it. Maybe we should add some relevant excerpts from migration documentation to the commit message of 037ec60e917dbdd7550e3741513b8b06cdada3e6 and not have next three commits refer to this one as **previous commit**: "Please check the upgrading from v5 to v6 guide of the **previous commit** message to understand why we changing the structure"

File: package.json
Comment (Pl217) [inline package.json]: Why was `@mui/styles` removed in 740e3d57d30afe8288dacbc00c961da58658b277? Also, `@nrwl/js` has been deprecated in favor of `@nx/js`, whereas `@nrwl/tao` has been deprecated in favor of `nx`.

File: package.json
Comment (Pl217) [inline package.json]: Looks like you don't need `@nx/web` package, which you actually stopped using when you upgraded to Nx 15 in 9f2cb06fb85e935208b60b75b7da3f1abbc44145

File: package.json
Comment (Pl217) [inline package.json]: Why do we need this `jest-environment-jsdom` package now? And why it isn't on v29 like rest of Jest packages?

File: jest.preset.js
Comment (Pl217) [inline jest.preset.js]: Use the new defaults for snapshot format, we don't have any snapshots in this repo anyway :) Just remove this config.

File: libs/hpc-core/.babelrc
Comment (Pl217) [inline libs/hpc-core/.babelrc]: Why did you keep `@nrwl/`? There are usages in other places too, so I won't comment everywhere. Search and replace them with `@nx/`

File: apps/hpc-cdm/src/app/components/form-assignments-list.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/components/form-assignments-list.tsx]: In 5d4eef5583ac76359279e582fad41a74323b0ee1, you have changes to two `webpack.js` files which are the only ones belonging there. Rest of the changes are fixes of mistakes from previous commits. Don't commit those separately, but go back, rebase and amend (change) the relevant commits where changes like this formatting belong.

As a useful rule of thumb, any time you write "and" in your commit message, you are likely doing something that doesn't belong in that commit :)

**Review by Pl217 — CHANGES_REQUESTED** [review]
Thanks for separating your work into commits! Means a ton while reviewing. I have reviewed your work by looking at commits and that's also how I commented. Don't know whether that will be obvious when I publish inline comments.

You still seem to need some git polishing, but you're getting there!

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

File: jest.preset.js
Comment (Onitoxan) [reply jest.preset.js]: Nx migration didn't want to change it, I changed it by myself now

File: package.json
Comment (Onitoxan) [reply package.json]: https://github.com/mui/material-ui/issues/32142#issuecomment-1090002923

That github post is related on the decision of removing `@mui/styles` and related to the `@nrwl` packages remaning, the NX migration didn't change them automatically and when I tried to do it manually the Nx version of these two doesn't seem to be in npm (even though Nx in their documentation they speak of nx/tao and nx/js)

File: libs/hpc-core/.babelrc
Comment (Onitoxan) [reply libs/hpc-core/.babelrc]: (commented in the prvious related issue about this)

File: jest.preset.js
Comment (Pl217) [reply jest.preset.js]: There is a reason why Nx migration didn't change it. We cannot use `import ... from` syntax even if we make this a Typescript file. If we need to keep `require()`, then, let's leave it as JS. Sorry for misleading you.

File: package.json
Comment (Pl217) [inline package.json]: When I requested you to split your work and do gradual upgrades of Nx, you did split into individual commits, which is exactly what I wanted. But, each Nx upgrade comes with a migration tool, that updates `package.json` with newer versions of software, but also creates `migrations.json`, that you run with `nx migrate --run-migrations`. Exact steps are listed in [the guide](https://nx.dev/core-features/automate-updating-dependencies).

However, when you split updates into individual commits, you only made changes to `package.json` versions and didn't do migrations to make actual changes in our codebase. You should have upgraded gradually version-by-version, but also should have migrated in same manner.

File: apps/hpc-cdm/src/app/components/form-assignments-list.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/components/form-assignments-list.tsx]: There is absolutely no need for any context consumers and this CSS class, you just use simple `Tooltip` from MUI and style with `componentsProps` on it.

File: apps/hpc-cdm/src/app/components/enketo/powered-by-footer.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/components/enketo/powered-by-footer.tsx]: I don't know why you changed Enketo and KoBoToolbox logos at all. I have done same upgrades and didn't need to touch anything.

File: apps/hpc-cdm/src/app/pages/operation-cluster-forms.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/pages/operation-cluster-forms.tsx]: First big complaint regarding how you migrated to React Router v6 is the fact that you hardcoded the paths. There are bunch of places where you did that, so this comment is about all of them. The paths were previously defined in `apps/hpc-cdm/src/app/paths.ts` file and there is no need to hardcode them now.

Yes, resolving paths has changed. Previously they were absolute and there was `exact` option, but now they are relative to their parent path. You should have kept the centralized location where paths are defined.

We do have better readability with paths hardcoded, but it's terrible if something changes, since paths are scattered around. We can also improve readability with docstrings or with return type definitions.

File: apps/hpc-cdm/src/app/pages/operation-cluster-forms.tsx
Comment (Pl217) [inline apps/hpc-cdm/src/app/pages/operation-cluster-forms.tsx]: Second big complaint regarding how you migrated to React Router v6 is that you have a bunch of patterns like this. We can no longer have functions inside `<Route>` elements where we could do pre-processing and verification of URL parameters, but you have done exact thing in all the places where this pattern occurs:

1. Add local "temporary" (you even named it `TemporalForm` in one place) component and process parameters with `useParams()` inside of it
2. Use this temporary component with `<React element>`, now that all elements should be passed via that prop

My biggest complaint is actually that you recognized this as a pattern and then copy pasted same structure all over the place. You named such construct a "temporal" (probably meant "temporary") component in one place - `TemporalForm`, but then seem to have realized why such construct is needed (to validate URL path params) and named it more descriptive `ValidateParams`. You have repeated the same pattern 5 times without moving this into a separate component!

From reading the documentation, the maintainers seem to want us to use `useParams()` in components directly, that is fine, but you shouldn't have introduced a temporary component. What you should have done, according to documentation is exactly what you did in `apps/hpc-cdm/src/app/pages/operation.tsx`. Just use `useParams()` is component directly, no temporary components needed. That way, instead of component's error being handled outside, component will be responsible for checking its own props.

But, like I said, we need to go one step further and have a component that will do this for us, in a generic manner.

File: apps/hpc-ftsadmin/project.json
Comment (Pl217) [inline apps/hpc-ftsadmin/project.json]: I realized here that our webpack config for FTS Admin isn't used at all. Not your fault, it is like that on `develop` now. It is defined like this in `workspace.json` currently, but it should be:
```suggestion
        "webpackConfig": "apps/hpc-ftsadmin/webpack.js"
```

File: package.json
Comment (Pl217) [inline package.json]: You have fixed your previous bad conflict resolutions during a rebase, when a bunch of packages were reverting changes from #373. Now, you only have one bad conflict resolution, because you've brought back `@types/dotenv` which got removed in #376.

**Review by Pl217 — CHANGES_REQUESTED** [review]
My main concerns are that you didn't do Nx migration in steps and that you haven't upgraded codebase to React Router v6 properly (we need generic component for validating params and centralized routes). More info inline.

**Review by Pl217 — APPROVED** [review]
Ok, I have pushed fixes that address all of my own comments from above. @Onitoxan I need you to take a look at my work and make sure I didn't break anything you've done. Especially routes, which were tricky.

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Comment by Onitoxan** [comment]
I have reviewed commit per commit and everything works fine to merge, I just have some observations to know why you decided to do certain decisions and also some suggestions!:

In commit 53c9f508a667fab8d46606218cd5f8529f89ea36 : I think we could delete this commit, also I think it doesn't help transitioning to react v6 (Even if the documentation says so) Because we are nesting the components taking them out of the `component` attribute that actually seems to be more fitting and easy to transition to v6 `element` attribute. Also, you modify `operation.tsx` in this commit that doesn't seem to belong here. I would have updated this file in previous commit were you are deleting the comment and rename the commit message, I think it would make more sense 🤔 

In commit 7ca98cc743ba0d80c4ffb9f10a260f071669f6ff I see you modify `es.json` and some of the translations seems awkward (It is normal) would you want me to revise that file and update some typos?

In commit 3da19483a17bedf2162c9c0421766688755c12a7 in the file `paths.tsx` You let a comment on top of the code, why didn't you put it on the commit description instead in the code itself?
 
In commit 7b867044ff12ec264f0b1f63199c29d714b0d7dd I don't understand what's the use of HMR, if you could tell me, or link me a page to read about it I would thank it.


Despite these comments I don't have blocking change requests, it's ready for a merge from my side 👍

**Review by Pl217 — CHANGES_REQUESTED** [review]
@Onitoxan, those are all valid concerns, thanks for your feedback. I am going to answer one-by-one.

> In commit [53c9f50](https://github.com/UN-OCHA/hpc-cdm/commit/53c9f508a667fab8d46606218cd5f8529f89ea36) : I think we could delete this commit, also I think it doesn't help transitioning to react v6 (Even if the documentation says so) Because we are nesting the components taking them out of the `component` attribute that actually seems to be more fitting and easy to transition to v6 `element` attribute. Also, you modify `operation.tsx` in this commit that doesn't seem to belong here. I would have updated this file in previous commit were you are deleting the comment and rename the commit message, I think it would make more sense thinking

You are completely right. This commit is there only because official docs are saying it's a smart thing to do when preparing to migrate from React Router v5 to v6. I was thinking about removing it myself, but since I got fooled by documentation and wasted my time on these useless changes, I decided to keep it to justify my time spent :smile: But, you're right, it is useless and I will get rid of it.

On the other hand, I don't agree that changes to `operation.tsx` should be moved to a commit where I remove commented out code. That commit is for cleanup and we should not mix concerns. Since I'm deleting commit and changes of `operation.tsx` need to be moved somewhere else anyway, its natural place is in same commit other components are also switched to use `useParams()` hook.

> In commit [7ca98cc](https://github.com/UN-OCHA/hpc-cdm/commit/7ca98cc743ba0d80c4ffb9f10a260f071669f6ff) I see you modify `es.json` and some of the translations seems awkward (It is normal) would you want me to revise that file and update some typos?

Sure, I forgot that we have a developer with powers greater than those of Google Translate - a native Spanish speaker :) I will use the messages which you shared with me on Slack.

> In commit [3da1948](https://github.com/UN-OCHA/hpc-cdm/commit/3da19483a17bedf2162c9c0421766688755c12a7) in the file `paths.tsx` You let a comment on top of the code, why didn't you put it on the commit description instead in the code itself?

Whereas you're partly right about this comment being slightly biased, meaning it reads well when viewed in a diff now, but would be slightly out of place once it becomes a legacy code.
However, I still want to convey the message why some routes must have leading/trailing slashes and why the code in that file is slightly weird. Thus, I will remove reference to React Router v5 in this comment. Also, since I want this message to be visible, it doesn't belong in commit message.

> In commit [7b86704](https://github.com/UN-OCHA/hpc-cdm/commit/7b867044ff12ec264f0b1f63199c29d714b0d7dd) I don't understand what's the use of HMR, if you could tell me, or link me a page to read about it I would thank it.

HMR stands for Hot Module Replacement (I actually thought R is for Reload), which is a [Webpack feature](https://webpack.js.org/concepts/hot-module-replacement/) that allows partial reloads of application.

> Despite these comments I don't have blocking change requests, it's ready for a merge from my side 👍

No, you've nailed a few very valid points here, so I will address them soon and then merge.

**Review by unocha-hpc — APPROVED** [review]
Checks have passed and this pull request is ready for manual review

**Review by Pl217 — APPROVED** [review]

**Pull request merged**
